book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+     pub hash: Hash,
++    #[solana_abi_added(since = "rev1")]
++    /// processing timestamp of last slot
++    pub timestamp: Option<UnixTimestamp>,
+ }
+
+ impl SolanaABI for Vote {
+   const REVISION_LIST = [
+      (0, "epoch", 0..), // leave this revision acceptable for indefinite time
++     (1, "epoch", 21..),
+   ]
+ }
+
+
+@@test code@@
+ #[test]
+ #[solana_abi_test]
+ test_serialize_vote() {
+    //   ^^^ no change is needed for ABI test case
+    let vote = ...
+    serializing_with_fixture!(vote, |write_stream, revision_tuple| {
+        // use revision_tuple to differentiate serialization in some way...
+        serialize_into(write_stream, normal_vote)
+    });
+    ...
+ }
+
+ #[test]
+ #[solana_abi_test]
+ test_deserialize_vote() {
+    //   ^^^ no change is needed for ABI test case
+    let vote = ...
+    deserializing_with_fixture!(Ok(vote), |read_stream, revision_tuple| {
+        // use revision_tuple to differentiate deserialization in some way..
+        deserialize_from(write_stream)
+    });
+ }
+
+Files:
+ test/fixtures/vote-rev-0.hex.yaml
++test/fixtures/vote-rev-1.hex.yaml
+```
+
+## Example 2: If we were adding #7415 after some time since mainnet launch:
+
+```patch
+ /// Meta contains enough context to recover the index from storage itself
+ /// This struct will be backed by mmaped and snapshotted data files.
+ /// So the data layout must be stable and consistent across the entire cluster!
+-#[solana_abi(allow=none, digest="662792c9", test="test_hash_stored_account")]
++#[solana_abi(allow=none, digest="c5031113", test="test_hash_stored_account")]
+ #[derive(PartialEq, Debug)]
+ pub struct StoredAccount<'a> {
+     pub meta: &'a StoredMeta,
+     /// account data
+     pub account_meta: &'a AccountMeta,
+     pub data: &'a [u8],
+     pub offset: usize,
++    #[solana_abi_changed(since = "rev1", comment="mix in rent_epoch and executable")
+     pub hash: &'a Hash,
+ }
+
+ impl SolanaABI for StoredAccount {
+   const REVISION_LIST = [
+-     (0, "cluster", "v0.1.1"..),
++     (1, "cluster", "v0.1.1".."v1.3.0"), // Reject/ignore after some time
++     (2, "cluster", "v1.2.0"..),
+   ]
+   # or
+   const REVISION_LIST = [
+-     (0, "snapshot_format", "1.."),
++     (1, "snapshot_format", "1"),
++     (2, "snapshot_format", "2.."),
+   ]
+
+  fn hash_account_data(
+      ...,
+      snapshot_format_version: u8,
+  ) -> Hash {
+        LittleEndian::write_u64(&mut buf[..], slot);
+        hasher.hash(&buf);
+
+        if (snapshot_format_version >= 2) { 
+            LittleEndian::write_u64(&mut buf[..], rent_epoch);
+            hasher.hash(&buf);
+        }
+
+        hasher.hash(&data);
+        ...
+   }
+
+   ...
+
+   // no change for ABI test is needed.
+Files:
+ .../fixtures/stored-account-rev-0.hex.yaml
++.../fixtures/stored-account-rev-1.hex.yaml
+```
+
+# Compatibility strictness
+
+(To be documented: table of compatibility of compatibility item types, status
+and kind of changes)
```

Comment:
Also, as for `stirctnness`, I have yet to answer to this original point from [the issue](#7421)...:

> general agreement on how strict we want to be. eg, once the system program is published/locked down, would it be permitted to even add a new system instruction? Most strictly, no

Reply:
This got out-of-scope, too.